mathesar/utils/permissions.py

```diff
@@ -1,20 +1,118 @@
     """Move data from old-style connection model to new models."""
     conn = Connection.current_objects.get(id=connection_id)
+    user = User.objects.get(id=user_id)
+    return _setup_connection_models(
+        conn.host, conn.port, conn.db_name, conn.username, conn.password, user
+    )
+
+
+@transaction.atomic
+def set_up_new_database_for_user_on_internal_server(
+        database_name, user, sample_data=[]
+):
+    """
+    Create a database on the internal server and install Mathesar.
+
+    This database will be set up to be accessible for the given user.
+    """
+    conn_info = settings.DATABASES[INTERNAL_DB_KEY]
+    if database_name == conn_info["NAME"]:
+        raise BadInstallationTarget(
+            "Mathesar can't be installed in the internal database."
+        )
+    user_database_role = _setup_connection_models(
+        conn_info["HOST"],
+        conn_info["PORT"],
+        database_name,
+        conn_info["USER"],
+        conn_info["PASSWORD"],
+        user
+    )
+    install_mathesar(
+        database_name,
+        conn_info["USER"],
+        conn_info["PASSWORD"],
+        conn_info["HOST"],
+        conn_info["PORT"],
+        True,
+        root_db=conn_info["NAME"],
+    )
+    with user_database_role.connection as conn:
+        _load_sample_data(conn, sample_data)
+    return user_database_role
+
 
-    server = Server.objects.get_or_create(host=conn.host, port=conn.port)[0]
-    database = Database.objects.get_or_create(name=conn.db_name, server=server)[0]
-    role = Role.objects.get_or_create(
-        name=conn.username, server=server, password=conn.password
-    )[0]
+@transaction.atomic
+def set_up_preexisting_database_for_user(
+        host, port, database_name, role_name, password, user, sample_data=[]
+):
+    internal_conn_info = settings.DATABASES[INTERNAL_DB_KEY]
+    if (
+            host == internal_conn_info["HOST"]
+            and port == internal_conn_info["PORT"]
+            and database_name == internal_conn_info["NAME"]
+    ):
+        raise BadInstallationTarget(
+            "Mathesar can't be installed in the internal database."
+        )
+    user_database_role = _setup_connection_models(
+        host, port, database_name, role_name, password, user
+    )
+    install_mathesar(
+        database_name, role_name, password, host, port, True, create_db=False,
+    )
+    with user_database_role.connection as conn:
+        _load_sample_data(conn, sample_data)
+    return user_database_role
+
+
+@transaction.atomic
+def _setup_connection_models(
+        host, port, database_name, role_name, password, user
+):
+    server, _ = Server.objects.get_or_create(host=host, port=port)
+    database, _ = Database.objects.get_or_create(
+        name=database_name, server=server
+    )
+    role, _ = Role.objects.get_or_create(
+        name=role_name,
+        server=server,
+        defaults={"password": password},
+    )
     return UserDatabaseRoleMap.objects.create(
```

Comment:
This throws up an `IntegrityError` when we try to add the existing credentials via the API, I think we ought to use `get_or_create` here as well, to ensure that we allow users to reinstall mathesar even if the database gets deleted outside of mathesar.

Reply:
I'm okay with that change.